### ISSUE TITLE
Question regarding the LSTM notations in the paper

### ISSUE THREAD

**Issue by Donglin-Wang2** [YicongHong/Fine-Grained-R2R#1]
In equation 1 on page 3363 of the paper, the LSTM has (uhat_t, m_t-1) as the second parameter. **I am wondering what the parenthesis does around the attended text feature (ATF) and the previous memory (PM).** Does it mean we are concatenating the ATF and PM into one vector, or adding them, or taking the inner product?

**Comment by YicongHong** [comment]
Hi Donglin, uhat_t and m_t-1 in Equation 1 are the hidden state and the cell state (memory) of the LSTM, respectively. Here we follow the implementation of the agent's state LSTM in [BackTranslation Agent](https://github.com/airsplay/R2R-EnvDrop/blob/c416108e1f713a5fea2e2bda887ab22080fb3cd1/r2r_src/model.py#L178) to replace the hidden state from previous time step (t-1) by the attended textual features at time step (t).

Thanks a lot for pointing this out, I didn't notice the problem with this equation --- I have to admit that Section 3.1 in the paper takes the [Self-Monitoring Agent](https://arxiv.org/abs/1901.03035) to present the information flow, but Equation 1 is for our fourth experiment with the [BackTranslation Agent](https://arxiv.org/abs/1904.04195). --- The corrected version should be LSTM([a_t-1; vhat_t; uhat_t], (h_t-1, m_t-1)) as in [this code](https://github.com/chihyaoma/selfmonitoring-agent/blob/b674ca6822e8282fafe327fea1d3c233f1956764/tasks/R2R-pano/models/policy_model.py#L88).

**Comment by IMNearth** [comment]
Hi Yicong, I share the same question with Donglin. Does your reply indicates that the sub-instruction aware agent described in your paper follows the structure of Self-Monitoring Agent ?

**Comment by YicongHong** [comment]
Hi there, yes, except Equation 1 as I explained above.

**Comment by IMNearth** [comment]
Thank you very much! Another question I have is about the Equation (5) in paper. As described , "v^a_t is the visual feature at the selected action direction", so what's the difference between v^a_t and a_t ? I think their dimensions are both 2176 ... Besides, since we "sample" actions during training, and those already ended data in a batch should not do actions anymore, how can I keep v^a_t 'requires_grad=True' ?

**Comment by YicongHong** [comment]
v^a_t is the visual feature at the selected action direction, dim 2176
a_t is the action (directional) encoding, dim 128

Sorry I didn't really get your second question. But I think v^a_t doesn't need to have gradient because the visual part (dim 2048) is directly from a pre-trained ResNet and the directional part (dim 128) is mannually assigned. On the other hand, it doesn't matter if you keep computing the ended agents, because usually people use the "ended" flag to prevent those stopped agents to move and use a "ignoreid" to ignore the loss of the stopped agents (hence no gradient).

**Comment by IMNearth** [comment]
Let me make my first question clearer. 
In Equation 1 as you mentioned above, the correct version is LSTM([a_t-1; vhat_t; uhat_t], (h_t-1, m_t-1)), where 'a_t-1' is the vector representing the previous action selected by the agent. In this [code](https://github.com/chihyaoma/selfmonitoring-agent/blob/b674ca6822e8282fafe327fea1d3c233f1956764/tasks/R2R-pano/models/policy_model.py#L88), it is a vector with  dim [batch, 2176] (before passed to mlp). So I might argue that 'v^a_t' and 'a_t' have the same meaning. 
However I have already found that this problem results from different notations in different paper. It's fine, I got it.

And thanks for your replying to the second question! Just because h_t and xhat_i are outputs of neural networks, so they have 'requires_grad=True'. Thus I assumed that v^a_t should also have gradient.

**Comment by IMNearth** [comment]
One more question is, I am not sure I understood this -- "We prevent the sub-instruction shifting module from converging to an undesirable local minimum by forcing the shifting loss to consider an equal number of randomly selected shift and do-not-shift samples in each time step." Could you please explain in detail about this operation ? How to 'force the loss to consider ...' and how to generate those 'shift and do-not-shift samples' ?

**Comment by YicongHong** [comment]
Good question. The motivation of this operation is in early training, the agent makes lots of mistakes so that it won't have the change to do lots of shifting -- most of the ground-truths for shifting in this case are "do not shift" -- which the agent might learn to keep choosing "do not shift" to minimize this loss, hence an undesirable local minimum.

The exact operation is, say we have a minibatch of size 64, the ground-truth for 10 of them is "shift" and the rest 54 is "do not shift", then we use all the 10 "shift" samples and randomly select 10 "do not shift" from that 54 samples to compute the shifting loss. In the opposite case, if "shift" is the majority, we simply do it the other way around -- by randomly choosing the same number of "shift" samples as the number of "do not shift".

And just a kind reminder, I have to say that the way we use the sub-instruction-path pairs in our paper is a very naive and not recommanded way, the model is nothing but a very simple baseline. An alternative solution could be to use the fine-grained annotations as soft guidance and allow the agent to benefit from both the local and global information at each step.

**Comment by IMNearth** [comment]
Thank you, Yicong, for your kind reply and remind. To be honest, I am trying to replicate the experiments in your paper so I am focusing the details... And I think I will try to use the fine-grained annotations as soft guidance (this is quite a natural idea) after the replication.
